### ISSUE TITLE
"Push Error" thrown for larger objects

### ISSUE THREAD

**Issue by CherryDT** [hildjj/node-cbor#68]
After a certain object size, encoding fails with the error "Push Error" which is meaningless. There is no limit shown in the docs (at least I didn't find it). How can I increase this limit?

```
const arr = []
for(let i = 0; i < 50000; i++) arr.push(1)
const result = require('cbor').encode(arr) // throws error
```

**Comment by hildjj** [comment]
Agree there should be a better error.  There are a couple of options, neither of which is as easy to use as `encode`.  Your issue is that the internal transform stream is reaching its highWaterMark, which defaults to 16384.  Either way, you'll want to not use the convenience function, but create an Encoder instance.

1) stream the output:

```js
const enc = new Encoder()
enc.pipe(process.stdout)
enc.write(arr)
```

2) set the highwaterMark for the stream:

```js
const e = new cbor.Encoder({highWaterMark: 1<<30})
enc.write(arr)
const result = enc.read()
```

The latter may feel easier to use, but will use unbounded amounts of memory on your system, which may be an issue, even a *security* issue, if you're parsing untrusted inputs.

**Comment by hildjj** [comment]
(leaving this open for a while in case someone else has a better idea)

**Comment by hildjj** [comment]
Added one small fix and docs for this in the fix to #93

**Comment by ahamid** [comment]
I am having a lot of trouble not inducing this error. My interpretation/intuition was that as long as the  appropriate highWaterMark is set to be larger than the largest single encoded object this problem can be avoided, but it doesn't seem to be the case: I can always reproduce it for a stream (series of writes) large enough. Curiously this problem is happening when writing to a file stream, but completely disappears when writing to stdout/stderr (despite those streams reporting the same `writableHighWaterMark`). I can seem to circumvent this issue by setting either of the encoder `readableHighWaterMark` or the output stream `highWaterMark` to larger than the total size of the incoming input - but that defeats the point of streaming.

Am I missing something very remedial with regard to node streams? Does backpressure need to be manually handled in this case (this is more complex than the sample stream solution above)?

```javascript
describe('Stream encoding', function() {
  this.timeout(60000);
  const obj_size = 2 * 1024;
  const objects = [];
  for (let i = 0; i < 100; i++) {
    objects.push(new Array(obj_size).fill(String.fromCharCode('a'.charCodeAt(0) + Math.floor(Math.random() * 26))).join(''));
  }
  it('should stream to file', async () => {
    // either encoder readableHighWaterMark or out stream highWaterMark
    // must be larger than the the entire encoded stream or encoder will eventually
    // throw Push Error after a resume; otherwise combinations of their relative values
    // don't seem to affect success
    // encoder readableHighWaterMark affects when write backpressure is induced
    const encHighWater = 30 * 1024; // * 1024 // success regardless of other settings
    const outHighWater = 20 * 1024; // * 1024 // success regardless of other settings
    // writableHighWaterMark doesn't seem to have much effect: if it's set to above 1 then there is no initial back pressure writing objects
    const enc = new Encoder({ readableHighWaterMark: encHighWater, writableHighWaterMark: 2 });
    return withTmp(async (file) => {
      // file stream highWaterMark induces encoder reading to pause
      const out = createWriteStream(file, { highWaterMark: outHighWater });
      // const out = process.stderr; // succeeds regardless of settings
      enc.pipe(out);
      for (let i = 0; i < objects.length; i++) {
        console.log('Writing object', i);
        try {
          if (!enc.write(objects[i])) {
            console.log('Backpressure at object ', i);
          }
        } catch (error) {
          console.log('Error writing', error); // Push Error after resume event
          throw error;
        }
      }
      enc.end();
      await finished(enc);
      await finished(out);
    });
  });
});
```
